mesh/mesh.hpp

```diff
@@ -2445,7 +2452,338 @@ class Mesh
+      The "local" group (with index 0) does not have any shared entities, so the
+      0-th row in the Table is always empty.
+
+      IMPORTANT: the descriptions of the groups in this MeshPart must match
+      their descriptions in all neighboring MeshParts. This includes the
+      ordering of the shared entities within the group, as well as the vertex
+      ordering of each shared entity.
+   */
+   Table group_shared_entity_to_vertex[Geometry::NumGeom];
+
+   ///@}
+
+   /** @brief Write the MeshPart to a stream using the parallel format
+       "MFEM mesh v1.2". */
+   void Print(std::ostream &os) const;
+
+   /** @brief Construct a serial Mesh object from the MeshPart.
+
+       The nodes of 'mesh' are NOT initialized by this method, however, the
+       nodal FE space and nodal GridFunction can be created and then attached to
+       the 'mesh'. The Mesh is constructed only if 'mesh' is empty, otherwise
+       the method simply returns the object held by 'mesh'.
+   */
+   Mesh &GetMesh();
+};
+
+
+/** @brief Class that allows serial meshes to be partitioned into MeshPart
+    objects, typically one MeshPart at a time, which can then be used to write
+    the local mesh in parallel MFEM mesh format.
+
+    Sample usage of this class: partition a serial mesh and save it in parallel
+    MFEM format:
+    \code
+       // The array 'partitioning' can be obtained e.g. from
+       // mesh->GeneratePartitioning():
+       void usage1(Mesh *mesh, int num_parts, int *partitioning)
+       {
+          MeshPartitioner partitioner(*mesh, num_parts, partitioning);
+          MeshPart mesh_part;
+          for (int i = 0; i < num_parts; i++)
+          {
+             partitioner.ExtractPart(i, mesh_part);
+             ofstream omesh(MakeParFilename("my-mesh.", i));
+             mesh_part.Print(omesh);
+          }
+       }
+    \endcode
+
+    This class can also be used to partition a mesh and GridFunction(s) and save
+    them in parallel:
+    \code
+       // The array 'partitioning' can be obtained e.g. from
+       // mesh->GeneratePartitioning():
+       void usage2(Mesh *mesh, int num_parts, int *partitioning,
+                   GridFunction *gf)
+       {
+          MeshPartitioner partitioner(*mesh, num_parts, partitioning);
+          MeshPart mesh_part;
+          for (int i = 0; i < num_parts; i++)
+          {
+             partitioner.ExtractPart(i, mesh_part);
+             ofstream omesh(MakeParFilename("my-mesh.", i));
+             mesh_part.Print(omesh);
+             auto lfes = partitioner.ExtractFESpace(mesh_part, *gf->FESpace());
+             auto lgf = partitioner.ExtractGridFunction(mesh_part, *gf, *lfes);
+             ofstream ofield(MakeParFilename("my-field.", i));
+             lgf->Save(ofield);
+          }
+       }
+    \endcode
+*/
+class MeshPartitioner
+{
+protected:
+   Mesh &mesh;
+   int *partitioning;
+   bool own_partitioning; // true == constructed with NULL partitioning
```

Comment:
Maybe this would be cleaner if `partitioning` was an `Array<int>`, which would mean we can get rid of `own_partitioning` and the destructor for this class.

```diff
diff --git a/mesh/mesh.cpp b/mesh/mesh.cpp
index f7caf761c..d832df5d7 100644
--- a/mesh/mesh.cpp
+++ b/mesh/mesh.cpp
@@ -13602,18 +13602,18 @@ MeshPartitioner::MeshPartitioner(Mesh &mesh_,
                                  int num_parts_,
                                  int *partitioning_,
                                  int part_method)
-   : mesh(mesh_),
-     partitioning(partitioning_),
-     own_partitioning(false)
+   : mesh(mesh_)
 {
-   if (partitioning == nullptr)
+   if (partitioning_)
    {
-      partitioning = mesh.GeneratePartitioning(num_parts_, part_method);
-      own_partitioning = true;
+      partitioning.MakeRef(partitioning_, mesh.GetNE(), false);
+   }
+   {
+      partitioning_ = mesh.GeneratePartitioning(num_parts_, part_method);
+      partitioning.MakeRef(partitioning_, mesh.GetNE(), true);
    }
 
-   Transpose(Array<int>(partitioning, mesh.GetNE()),
-             part_to_element, num_parts_);
+   Transpose(partitioning, part_to_element, num_parts_);
    // Note: the element ids in each row of 'part_to_element' are sorted.
 
    const int dim = mesh.Dimension();
@@ -14174,11 +14174,6 @@ MeshPartitioner::ExtractGridFunction(const MeshPart &mesh_part,
    return local_gf;
 }
 
-MeshPartitioner::~MeshPartitioner()
-{
-   if (own_partitioning) { delete [] partitioning; }
-}
-
 
 GeometricFactors::GeometricFactors(const Mesh *mesh, const IntegrationRule &ir,
                                    int flags, MemoryType d_mt)
diff --git a/mesh/mesh.hpp b/mesh/mesh.hpp
index 17b6cdde4..1114794e0 100644
--- a/mesh/mesh.hpp
+++ b/mesh/mesh.hpp
@@ -2707,8 +2707,7 @@ class MeshPartitioner
 {
 protected:
    Mesh &mesh;
-   int *partitioning;
-   bool own_partitioning; // true == constructed with NULL partitioning
+   Array<int> partitioning;
    Table part_to_element;
    Table part_to_boundary;
    Table edge_to_element;
@@ -2780,9 +2779,6 @@ public:
    ExtractGridFunction(const MeshPart &mesh_part,
                        const GridFunction &global_gf,
                        FiniteElementSpace &local_fespace) const;
-
-   /// Destructor.
-   ~MeshPartitioner();
 };
 ```

Reply:
Done in 63721b08e74010f7a28ac6c356aa1e8aba62f850 and 1f39aba374f6323a402403e853bbd8c5bc304cc6